alf/algorithms/causal_bc_algorithm.py

```diff
@@ -0,0 +1,210 @@
+            reward_weights=None,
+            env=env,
+            config=config,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        self._actor_network = actor_network
+        self._discriminator_network = discriminator_network
+
+        if actor_optimizer is not None and actor_network is not None:
+            self.add_optimizer(actor_optimizer, [actor_network])
+        self._actor_optimizer = actor_optimizer
+
+        if discriminator_optimizer is not None and discriminator_network is not None:
+            self.add_optimizer(discriminator_optimizer,
+                               [discriminator_network])
+        self._discriminator_optimizer = discriminator_optimizer
+
+        self._bc_regulatization_weight = bc_regulatization_weight
+        self._f_norm_penalty_weight = f_norm_penalty_weight
+
+    def _predict_action(self, observation, state):
+        action_dist, actor_network_state = self._actor_network(
+            observation, state=state)
+
+        return action_dist, actor_network_state
+
+    def predict_step(self, inputs: TimeStep, state: BcState):
+        action_dist, new_state = self._predict_action(
+            inputs.observation, state=state.actor)
+        action = dist_utils.epsilon_greedy_sample(action_dist,
+                                                  self._epsilon_greedy)
+
+        return AlgStep(output=action, state=BcState(actor=new_state))
+
+    def residuIL_loss(self, targets, predictions, pred_residuals):
+
+        # train policy (detach discriminator)
+        policy_loss = (2 * (targets - predictions) * pred_residuals.detach()
+                       ).mean(-1) + self._bc_regulatization_weight * (
+                           torch.square(targets - predictions)).mean(-1)
+
+        # train discriminator (detach policy)
+        discriminator_loss = -(
+            2 * (targets - predictions).detach() * pred_residuals -
+            pred_residuals * pred_residuals).mean(-1)
+
+        discriminator_loss = (
+            discriminator_loss +
+            self._f_norm_penalty_weight * torch.linalg.norm(pred_residuals))
+
+        return policy_loss, discriminator_loss
+
+    def train_step_offline(self,
+                           inputs: TimeStep,
+                           state,
+                           rollout_info,
+                           pre_train=False):
+
+        action_dist, new_state = self._predict_action(
+            inputs.observation, state=state.actor)
+
+        predictions = dist_utils.get_rmode(action_dist)
+        pred_residuals, _ = self._discriminator_network(inputs.observation)
+
+        info = BcInfo(
+            actor=predictions,
+            discriminator=pred_residuals,
+            target=rollout_info.action)
+        return AlgStep(
+            rollout_info.action, state=BcState(actor=new_state), info=info)
+
+    def calc_loss_offline(self, info, pre_train=False):
+
+        #[T, B, 1]
```

Comment:
"1" should be the dimension of action?

Reply:
Yes, should be action dimension. Updated.